service/storage_service.hh

```diff
@@ -295,9 +295,11 @@ private:
         locator::endpoint_dc_rack dc_rack;
         locator::host_id host_id;
     };
-    future<replacement_info> prepare_replacement_info(std::unordered_set<gms::inet_address> initial_contact_nodes,
+    future<> prepare_replacement_info(std::unordered_set<gms::inet_address> initial_contact_nodes,
             const std::unordered_map<gms::inet_address, sstring>& loaded_peer_features);
 
+    std::optional<replacement_info> _replacement_info;
```

Comment:
I won't object to this, but I don't like putting more state inside the class :( It's not clear if this state is mutable, where it's needed etc. Would be nice if we could avoid it by just passing the `replacement_info` where we need it.

Reply:
I agree in principle, but unfortunately, the join_token_ring, handle_state_replacing song and dance requires this state :-(

Reply:
not present in v5, instead
the local replacement_info prepared in join_token_ring is used to pass the replace address down to functions that need it and are called from join_token_ring